harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,352 @@
+
+    def disable_dataset_reproducibility_checks(self) -> None:
+        self._data_repro_checks_disabled = True
+
+    @property
+    def use_pipeline_parallel(self) -> bool:
+        return self._use_pipeline_parallel
+
+    def _calculate_batch_sizes(self) -> Tuple[int, int]:
+        """Recompute per slot batch size to account for deepspeed support for micro-batches.
+
+        This needs to be done after the user calls wrap_model_engine to let us know the deepspeed
+        config which contains the micro-batch-size.
+        """
+        global_batch_size = self._global_batch_size
+
+        # Configure batch sizes.
+        num_replicas = self._mpu.get_data_parallel_world_size()
+
+        per_gpu_batch_size = global_batch_size // num_replicas
+        self._num_micro_batches_per_slot = per_gpu_batch_size // self.train_micro_batch_size_per_gpu
+        per_gpu_batch_size = self.num_micro_batches_per_slot * self.train_micro_batch_size_per_gpu
+        effective_batch_size = per_gpu_batch_size * num_replicas
+        if effective_batch_size != global_batch_size:
+            logging.warning(
+                f"`global_batch_size` changed from {global_batch_size} to {effective_batch_size} "
+                f"to account for deepspeed micro batch size."
+            )
+        return per_gpu_batch_size, effective_batch_size
+
+    @property
+    def train_micro_batch_size_per_gpu(self) -> int:
+        if self._train_micro_batch_size_per_gpu is None:
+            raise det.errors.InvalidExperimentException(
+                "Please call wrap_model_engine before accessing train_micro_batch_size."
+            )
+        return self._train_micro_batch_size_per_gpu
+
+    @property
+    def num_micro_batches_per_slot(self) -> int:
+        if self._num_micro_batches_per_slot is None:
+            raise det.errors.InvalidExperimentException(
+                "Please call wrap_model_engine before accessing num_micro_batches_per_slot."
+            )
+        return self._num_micro_batches_per_slot
+
+    def _set_determined_profiler(self, prof: profiler.ProfilerAgent) -> None:
+        self._determined_profiler = prof
+
+    @contextlib.contextmanager
+    def _record_timing(self, metric_name: str, accumulate: bool = False) -> Iterator[None]:
+        if not self._determined_profiler:
+            yield
+            return
+        with self._determined_profiler.record_timing(metric_name, accumulate):
+            yield
+
+    def _init_device(self) -> None:
+        self.n_gpus = len(self.env.container_gpus)
+        if not self.n_gpus:
+            raise det.errors.InvalidExperimentException("GPUs required for DeepSpeedTrial.")
+        if self.distributed.size > 1:
+            # We launch a separate process per GPU with LOCAL_RANK set by DeepSpeed's launcher.
+            # Each process needs to bind to a unique GPU.
+            self.device = torch.device("cuda", int(cast(str, os.environ.get("LOCAL_RANK"))))
```

Comment:
Why are your reading from the environment here instead of from the distributed context?